src/solvers/dgsem_structured/dg_2d.jl

```diff
@@ -195,6 +197,62 @@ end
   end
 end
 
+# TODO: nonconservative terms, remove dispatch on volume_flux used for dispatch
+#       for development
+@inline function split_form_kernel!(du::AbstractArray{<:Any,4}, u,
+                                    element, mesh::Union{StructuredMesh{2}, UnstructuredMesh2D},
+                                    nonconservative_terms::Val{true}, equations,
+                                    volume_flux::Tuple{Any,Any}, dg::DGSEM, cache, alpha=true)
+  @unpack derivative_split = dg.basis
+  @unpack contravariant_vectors = cache.elements
+  symmetric_flux, nonconservative_flux = volume_flux
+
+  # Apply the symmetric flux as usual
+  split_form_kernel!(du, u, element, mesh, Val(false), equations, symmetric_flux, dg, cache, alpha)
+
+  # Calculate the remaining volume terms using the nonsymmetric generalized flux
+  for j in eachnode(dg), i in eachnode(dg)
+    u_node = get_node_vars(u, equations, dg, i, j, element)
+
+    # pull the contravariant vectors in each coordinate direction
+    Ja1_node = get_contravariant_vector(1, contravariant_vectors, i, j, element)
+    Ja2_node = get_contravariant_vector(2, contravariant_vectors, i, j, element)
+
+    # The diagonal terms are zero since the diagonal of `derivative_split`
+    # is zero. We ignore this for now.
+    # In general, nonconservative fluxes can depend on both the contravariant
+    # vectors (normal direction) at the current node and the averaged ones.
+    # Thus, we need to pass both to the nonconservative flux.
+
+    # x direction
+    integral_contribution = zero(u_node)
+    for ii in eachnode(dg)
+      u_node_ii = get_node_vars(u, equations, dg, ii, j, element)
+      # pull the contravariant vectors and compute the average
+      Ja1_node_ii = get_contravariant_vector(1, contravariant_vectors, ii, j, element)
+      Ja1_avg = 0.5 * (Ja1_node + Ja1_node_ii)
+      # Compute the contravariant nonconservative flux.
+      fluxtilde1 = nonconservative_flux(u_node, u_node_ii, Ja1_node, Ja1_avg, equations)
+      integral_contribution = integral_contribution + derivative_split[i, ii] * fluxtilde1
+    end
+
+    # y direction
+    for jj in eachnode(dg)
+      u_node_jj = get_node_vars(u, equations, dg, i, jj, element)
+      # pull the contravariant vectors and compute the average
+      Ja2_node_jj = get_contravariant_vector(2, contravariant_vectors, i, jj, element)
+      Ja2_avg = 0.5 * (Ja2_node + Ja2_node_jj)
+      # compute the contravariant nonconservative flux in the direction of the
+      # averaged contravariant vector
+      fluxtilde2 = nonconservative_flux(u_node, u_node_jj, Ja2_node, Ja2_avg, equations)
+      integral_contribution = integral_contribution + derivative_split[j, jj] * fluxtilde2
+    end
+
+    # The factor 0.5 cancels the factor 2 in the flux differencing form
```

Comment:
Aha, so the factor of 1/2 got factored out of the Powell flux and lives here now.

Reply:
Yes, exactly. That's 
- conceptually cleaner (in my opinion) since you don't need to know anything about our performance hacks of storing `2 * D` etc.
- computationally more efficient